File: simulator/src/java/main/ctu/nengoros/comm/nodeFactory/demo/turtle/OneTurtle.java
package ctu.nengoros.comm.nodeFactory.demo.turtle;

import ctu.nengoros.comm.nodeFactory.NodeGroup;
import ctu.nengoros.comm.rosutils.Mess;
import ctu.nengoros.comm.rosutils.RosUtils;

/**
 * Runs demo with turtle, where turtle has its own controlled and sensoric system 
 * which registers position of turtle in the map.
 * 
 * 
 * This demo launches a group of nodes consisting of 2 java nodes and one native
 * node, which is simulator of turtle, written in C++, part of ROS installation. 
 *
 * After pressing any key, all threads and processes should be ended.
 * 
 * Summary: 
 * 	-demo automatically starts roscore and rxgraph (if ROS installed), for each:
 * 		-process itself
 * 		-2 threads for consuming stdout and stderr
 * 		-1 thread checking whether to stop the process
 * -turtlesim process:
 * 		-process itself
 * 		-2 threads for consuming ..
 * 		-1 thread checking whether to stop
 * -controller and sensor, for each:
 * 		-1 thread with the NodeMain (ROS implementation of Node)
 * 	
 * summary:
 * 		-11 threads + the main one
 * 	
 * 
 * @author Jaroslav Vitku
 *
 */
public class OneTurtle {

	public static void main(String[] args) {
		
		RosUtils.prefferJroscore(false);

		String turtlesim = 	"../testnodes/turtlesim/turtlesim_node";
		String className = 	"resender.turtle.Controller";
		String clName = 	"resender.turtle.PositionSensor";

		// create group with a name
		NodeGroup g = new NodeGroup("zelvicka", true);
		
		// add node configurations to the group
		g.addNode(turtlesim, "zelva", "native");
		g.addNode(className, "actuators", "java");
		g.addNode(clName, "sensors", "java");
		
		// start all nodes in the group
		g.startGroup();
		
		Mess.waitForKey();
		
		// stop group
		g.stopGroup();
		
		RosUtils.utilsShallStop();
	}
}

File: simulator-ui/scripts/unsupported/other/oneTurtle.py
# Demo showing how the nengoros is able to launch group of nodes consisting of:
#   -java node actuators    = java node publishing commands by means of ROS network
#   -java node sensors      = java node modem subscribed to messages about turtle position
#   -native node            = native application (installed turtlesim, demo simulator shipped with ROS)

# as a result, (native application) turtle is controlled by java node actuators and sends messages to java node sensors
# all these nodes are in one nodeGroup

# note: that turtle application may run probably only on Ubuntu 12.04 64bit, ROS installation recommended
# note: that this is direct equivalent to Java class: nengoros.comm.nodeFactory.demo.turtle.OneTurtle
 
# by Jaroslav Vitku

import nef
import math
from ca.nengo.math.impl import FourierFunction
from ca.nengo.model.impl import FunctionInput
from ca.nengo.model import Units

# creates python class and calls constructor __init__
net=nef.Network('Random Turtle Control Demo')
net.add_to_nengo()  # here: delete old (toplevel) network and replace it with the newly CREATED one

# some default imports copied from nef
from ca.nengo.util import VisiblyMutableUtils
import java
import inspect
import warnings
# import my custom classes here:
from nengoros.modules.impl import DefaultNeuralModule as NeuralModule
from nengoros.comm.nodeFactory import NodeGroup as NodeGroup
from nengoros.comm.rosutils import RosUtils as RosUtils

################################## 
################# setup the ROS utils (optional) 
#RosUtils.setAutorun(False)  # Do we want to autorun roscore and rxgraph? (tru by default)
RosUtils.prefferJroscore(False) # Turlte prefers roscore before jroscore (don't know why..) 

################################## 
################# define the group 

turtlesim = "../testnodes/turtlesim/turtlesim_node"  # turtle under this project
act = "resender.turtle.Controller";
sense = "resender.turtle.PositionSensor";

# create group with a name
g = NodeGroup("zelvicka", True);
		
# add node configurations to the group
g.addNC(turtlesim, "zelva", "native");  # start native node called zelva
g.addNC(act, "actuators", "java");      # start java nactuator ndoe called actuators
g.addNC(sense, "sensors", "java");      # start java sensoric node called sensors

################################## 
################# start the group 
g.startGroup()

# that is it, this will run forever.. 
print "OK, configuration done."
print "Note that this demo can be stopped only by closing the Nengo simulator window"




